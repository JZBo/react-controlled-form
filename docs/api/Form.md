# `<Form>`

The Form component is used to declare a new form context. Every form field within it's children is automatically assigned it.

## Props
| Prop | Type | Description |
| --- | --- | --- |
| formId | (*string*) | A unique identifier which is used to save the whole form data in our Redux store. |
| initialFields | (*Object?*) | Defines initial field data. Every field value is optional. |
| validate | (*Function?*) | A function to perform additional form validation.<br>It's first parameter matches the [callback shape](#callback-shape). |
| onChange | (*Function?*) | A function that is triggered on **every** change.<br>It's first parameter matches the [callback shape](#callback-shape). It additionally also receives `previousData` which contains the field data before the change happened. |
| onSubmit | (*Function?*) | A function that is triggered if the form gets submitted.<br>It's first parameter matches the [callback shape](#callback-shape). |

#### Callback Shape
```
FieldData = {
  value: any,
  isDisabled: boolean,
  isRequired: boolean,
  isValid: boolean
}

CallbackShape = {
  updateField: Function(fieldId: string, data: FieldData),
  data: { [fieldId: string]: FieldData }
}
```

## Example
```javascript
import { Form } from 'react-controlled-form'

const initialFields = {
  country: {
    isRequired: true,
    isValid: false
  },
  language: {
    isEnabled: false
  }
}

function onChange({ updateField, data }) {
  // enable the language field and set 'English' as default
  // as soon as country has been set
  if (data.country.isTouched && data.country.value) {
    updateField('language', {
      value: 'English',
      isEnabled: true
    })
  }
}

export default () => (
  <Form formId="contact" initialFields={initialFields} onChange={onChange}>
    /* form fields */
  </Form>
)
```
