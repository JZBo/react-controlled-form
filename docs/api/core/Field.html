
<!DOCTYPE HTML>
<html lang="en" >
    <head>
        <title>4.1.2. Field Â· Controlled Forms for React and Redux</title>
        <meta charset="UTF-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.1.0">
        <meta name="author" content="Robin Frischmann <robin@rofrischmann.de>">
        
        
    
    <link rel="stylesheet" href="../../../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../../../gitbook/gitbook-plugin-prism/prism.css">
                
            
                
                <link rel="stylesheet" href="../../../gitbook/gitbook-plugin-advanced-emoji/emoji-website.css">
                
            
                
                <link rel="stylesheet" href="../../../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../../../gitbook/gitbook-plugin-fontsettings/website.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../../../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../../../gitbook/images/favicon.ico" type="image/x-icon">

    
    
    <link rel="prev" href="Form.html" />
    

    </head>
    <body>
        
<div class="book">
    <div class="book-summary">
        
            
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

            
                <nav role="navigation">
                


<ul class="summary">
    
    

    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../../../">
            
                <a href="../../../">
            
                    
                    Read Me
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="../../Introduction.html">
            
                <a href="../../Introduction.html">
            
                    
                    1. Introduction
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.2.1" data-path="../../introduction/Motivation.html">
            
                <a href="../../introduction/Motivation.html">
            
                    
                    1.1. Motivation
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.2" data-path="../../introduction/Migration.html">
            
                <a href="../../introduction/Migration.html">
            
                    
                    1.2. Migration
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.3" data-path="../../introduction/Examples.html">
            
                <a href="../../introduction/Examples.html">
            
                    
                    1.3. Examples
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="../../Basics.html">
            
                <a href="../../Basics.html">
            
                    
                    2. Basics
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.3.1" data-path="../../basics/Forms.html">
            
                <a href="../../basics/Forms.html">
            
                    
                    2.1. Forms
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.2" data-path="../../basics/Fields.html">
            
                <a href="../../basics/Fields.html">
            
                    
                    2.2. Fields
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.3" data-path="../../basics/State.html">
            
                <a href="../../basics/State.html">
            
                    
                    2.3. State
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.4" data-path="../../Guides.html">
            
                <a href="../../Guides.html">
            
                    
                    3. Guides
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.4.1" data-path="../../guides/StoreSetup.html">
            
                <a href="../../guides/StoreSetup.html">
            
                    
                    3.1. Store Setup
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.2" data-path="../../guides/FormInitialization.html">
            
                <a href="../../guides/FormInitialization.html">
            
                    
                    3.2. Form Initialization
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.3" data-path="../../guides/UpdatingFields.html">
            
                <a href="../../guides/UpdatingFields.html">
            
                    
                    3.3. Updating Fields
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.4" data-path="../../guides/FormValidation.html">
            
                <a href="../../guides/FormValidation.html">
            
                    
                    3.4. Form Validation
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.5" data-path="../../guides/ValidationErrors.html">
            
                <a href="../../guides/ValidationErrors.html">
            
                    
                    3.5. Validation Error
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.6" data-path="../../guides/HandlingSubmit.html">
            
                <a href="../../guides/HandlingSubmit.html">
            
                    
                    3.6. Handling Submit
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.7" data-path="../../guides/UsingFormState.html">
            
                <a href="../../guides/UsingFormState.html">
            
                    
                    3.7. Using Form State
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.8" data-path="../../guides/ThirdPartyComponent.html">
            
                <a href="../../guides/ThirdPartyComponent.html">
            
                    
                    3.8. Using Third-Party Components
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.9" data-path="../../guides/DebuggingForms.html">
            
                <a href="../../guides/DebuggingForms.html">
            
                    
                    3.9. Debugging Forms
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.5" data-path="../../API.html">
            
                <a href="../../API.html">
            
                    
                    4. API Reference
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.5.1" >
            
                <span>
            
                    
                    4.1. Core
            
                </span>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.5.1.1" data-path="Form.html">
            
                <a href="Form.html">
            
                    
                    4.1.1. Form
            
                </a>
            

            
        </li>
    
        <li class="chapter active" data-level="1.5.1.2" data-path="Field.html">
            
                <a href="Field.html">
            
                    
                    4.1.2. Field
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.5.2" >
            
                <span>
            
                    
                    4.2. Utilities
            
                </span>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.5.2.1" data-path="../utils/mapDataToValues.html">
            
                <a href="../utils/mapDataToValues.html">
            
                    
                    4.2.1. mapDataToValues
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.2.2" data-path="../utils/validateWithRequired.html">
            
                <a href="../utils/validateWithRequired.html">
            
                    
                    4.2.2. validateWithRequired
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    

            </ul>
            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


                </nav>
            
        
    </div>

    <div class="book-body">
        
            <div class="body-inner">
                
                    

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href="../../.." >4.1.2. Field</a>
    </h1>
</div>




                    <div class="page-wrapper" tabindex="-1" role="main">
                        <div class="page-inner">
                            
<div id="book-search-results">
    <div class="search-noresults">
    
                                <section class="normal markdown-section">
                                
                                <h1 id="field">Field</h1>
<p>Field is the most common component. <br>It is used to render fully integrated form fields that ships all required methods and data.</p>
<h2 id="props">Props</h2>
<table>
<thead>
<tr>
<th>Prop</th>
<th>Type</th>
<th>Default</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>fieldId</td>
<td>(<em>string</em>)</td>
<td></td>
<td>A (unique) identifier used to store the field data in our Redux store.</td>
</tr>
<tr>
<td>initialData</td>
<td>(<em>Object?</em>)</td>
<td></td>
<td>Default field data that is used to initialize the field. The data might be overwritten by <code>initialFields</code> set by the wrapping <a href="Form.html">Form</a> component.</td>
</tr>
<tr>
<td>render</td>
<td>(<em>Function</em>)</td>
<td></td>
<td>A render function that renders the component shape. <br>It&apos;s first parameter is an object containing the following properties.</td>
</tr>
</tbody>
</table>
<h3 id="render-properties">Render Properties</h3>
<table>
<thead>
<tr>
<th>Property</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>value</td>
<td>(<em>any</em>)</td>
<td>The controlled field value.</td>
</tr>
<tr>
<td>isEnabled</td>
<td>(<em>boolean</em>)</td>
<td>Indicates if the field is enabled.</td>
</tr>
<tr>
<td>isRequired</td>
<td>(<em>boolean</em>)</td>
<td>Indicates if the field is required.</td>
</tr>
<tr>
<td>isValid</td>
<td>(<em>boolean</em>)</td>
<td>Indicates if the field is valid.</td>
</tr>
<tr>
<td>isTouched</td>
<td>(<em>boolean</em>)</td>
<td>Indicates if the field is touched.<br><em>(Yields true as soon as the very first user input happens, but remains false for initialization)</em>.</td>
</tr>
<tr>
<td>state</td>
<td>(<em>Object</em>)</td>
<td>The form-scoped state</td>
</tr>
<tr>
<td>formId</td>
<td>(<em>string</em>)</td>
<td>The formId passed to the wrapping <a href="Form.html">Form</a> component.</td>
</tr>
<tr>
<td>fieldId</td>
<td>(<em>string</em>)</td>
<td>The fieldId passed to the component.</td>
</tr>
<tr>
<td>isFormValid</td>
<td>(<em>boolean</em>)</td>
<td>Indicates whether the whole form is valid or not.</td>
</tr>
<tr>
<td>updateField</td>
<td>(<em>Function</em>)</td>
<td>Updates the field data.<br>Accepts an object with <code>isEnabled</code>, <code>isRequired</code>, <code>isTouched</code>, <code>isValid</code> and <code>value</code>. Each value is optional though.</td>
</tr>
<tr>
<td>updateState</td>
<td>(<em>Function</em>)</td>
<td>Updates the form-scoped state.<br>Takes an object of new state keys and merges those with the existing state.</td>
</tr>
</tbody>
</table>
<h2 id="example">Example</h2>
<pre class="language-"><code class="lang-javascript"><span class="token keyword">import</span> <span class="token punctuation">{</span> Field <span class="token punctuation">}</span> <span class="token keyword">from</span> <span class="token string">&apos;react-controlled-form&apos;</span>

<span class="token keyword">function</span> <span class="token function">Input</span><span class="token punctuation">(</span><span class="token punctuation">{</span> value<span class="token punctuation">,</span> updateField <span class="token punctuation">}</span><span class="token punctuation">)</span> <span class="token punctuation">{</span>
  <span class="token keyword">const</span> <span class="token function-variable function">onChange</span> <span class="token operator">=</span> e <span class="token operator">=&gt;</span> <span class="token function">updateField</span><span class="token punctuation">(</span><span class="token punctuation">{</span> value<span class="token punctuation">:</span> e<span class="token punctuation">.</span>target<span class="token punctuation">.</span>value <span class="token punctuation">}</span><span class="token punctuation">)</span>

  <span class="token keyword">return</span> <span class="token punctuation">(</span>
    <span class="token operator">&lt;</span>input value<span class="token operator">=</span><span class="token punctuation">{</span>value<span class="token punctuation">}</span> onChange<span class="token operator">=</span><span class="token punctuation">{</span>onChange<span class="token punctuation">}</span> <span class="token operator">/</span><span class="token operator">&gt;</span>
  <span class="token punctuation">)</span>
<span class="token punctuation">}</span>

<span class="token comment">// Usage</span>
<span class="token operator">&lt;</span>Field fieldId<span class="token operator">=</span><span class="token string">&quot;firstname&quot;</span> render<span class="token operator">=</span><span class="token punctuation">{</span>Input<span class="token punctuation">}</span> <span class="token operator">/</span><span class="token operator">&gt;</span>
<span class="token operator">&lt;</span>Field fieldId<span class="token operator">=</span><span class="token string">&quot;lastname&quot;</span> render<span class="token operator">=</span><span class="token punctuation">{</span>Input<span class="token punctuation">}</span> <span class="token operator">/</span><span class="token operator">&gt;</span>
</code></pre>
<h2 id="tips--tricks">Tips &amp; Tricks</h2>
<ul>
<li>To provide maximum accessibility, it is recommended to utilize the default DOM attributes such as <code>disabled</code> and <code>required</code>. We may use <code>isEnabled</code> and <code>isRequired</code> to set those attributes respectively.</li>
<li>Using <code>onInput</code> instead of <code>onChange</code>, <code>onKeyPress</code> or <code>onKeyDown</code> ensures that the event has the most recent value.</li>
<li>Field validation can easily be done within <code>onChange</code>. We can then set <code>isValid</code> within <code>updateField</code> to update the state. Altering the visual appearance depending on <code>isValid</code> helps to tell the user if the field is actually valid or not. Additional, we could show an error message if it&apos;s invalid.</li>
</ul>

                                
                                </section>
                            
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

                        </div>
                    </div>
                
            </div>

            
                
                <a href="Form.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page: 4.1.1. Form">
                    <i class="fa fa-angle-left"></i>
                </a>
                
                
            
        
    </div>

    <script>
        var gitbook = gitbook || [];
        gitbook.push(function() {
            gitbook.page.hasChanged({"page":{"title":"4.1.2. Field","level":"1.5.1.2","depth":3,"next":{"title":"4.2. Utilities","level":"1.5.2","depth":2,"ref":"","articles":[{"title":"4.2.1. mapDataToValues","level":"1.5.2.1","depth":3,"path":"docs/api/utils/mapDataToValues.md","ref":"docs/api/utils/mapDataToValues.md","articles":[]},{"title":"4.2.2. validateWithRequired","level":"1.5.2.2","depth":3,"path":"docs/api/utils/validateWithRequired.md","ref":"docs/api/utils/validateWithRequired.md","articles":[]}]},"previous":{"title":"4.1.1. Form","level":"1.5.1.1","depth":3,"path":"docs/api/core/Form.md","ref":"docs/api/core/Form.md","articles":[]},"dir":"ltr"},"config":{"plugins":["edit-link","prism","-highlight","github","anker-enable","advanced-emoji"],"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"pluginsConfig":{"prism":{},"github":{"url":"https://github.com/rofrischmann/react-controlled-form/"},"anker-enable":{},"search":{},"lunr":{"maxIndexSize":1000000},"fontsettings":{"theme":"white","family":"sans","size":2},"advanced-emoji":{"embedEmojis":false},"sharing":{"facebook":true,"twitter":true,"google":false,"weibo":false,"instapaper":false,"vk":false,"all":["facebook","google","twitter","weibo","instapaper"]},"edit-link":{"base":"https://github.com/rofrischmann/react-controlled-form/tree/master","label":"Edit This Page"},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false}},"theme":"default","author":"Robin Frischmann <robin@rofrischmann.de>","pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"variables":{},"title":"Controlled Forms for React and Redux","language":"en","gitbook":"3.1.0","description":"Official documentation for react-controlled-form"},"file":{"path":"docs/api/core/Field.md","mtime":"2018-04-10T07:50:40.000Z","type":"markdown"},"gitbook":{"version":"3.1.0","time":"2018-08-09T17:17:57.879Z"},"basePath":"../../..","book":{"language":""}});
        });
    </script>
</div>

        
    <script src="../../../gitbook/gitbook.js"></script>
    <script src="../../../gitbook/theme.js"></script>
    
        
        <script src="../../../gitbook/gitbook-plugin-edit-link/plugin.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-github/plugin.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-anker-enable/anker.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-sharing/buttons.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-fontsettings/fontsettings.js"></script>
        
    

    </body>
</html>

