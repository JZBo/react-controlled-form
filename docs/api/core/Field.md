# Field

Field is the most common component. <br>It is used to render fully integrated form fields that ships all required methods and data.

## Props
| Prop | Type | Default | Description |
| --- | --- | --- | --- |
| fieldId | (*string*) | | A (unique) identifier used to store the field data in our Redux store. |
| initialData | (*Object?*) | | Default field data that is used to initialize the field. The data might be overwritten by `initialFields` set by the wrapping [`<Form>`](Form.md) component. | 
| render | (*Function*) | | A render function that renders the component shape. <br>It's first parameter is an object containing the following properties. |

### Render Properties

| Property | Type | Description |
| --- | --- | --- |
| value | (*any*) | The controlled field value. |
| isEnabled | (*boolean*) | Indicates if the field is enabled. |
| isRequired | (*boolean*) | Indicates if the field is required. |
| isValid | (*boolean*) | Indicates if the field is valid. |
| isTouched | (*boolean*) | Indicates if the field is touched.<br>*(Yields true as soon as the very first user input happens, but remains false for initialization)*. |
| state | (*Object*) | The form-scoped state |
| formId | (*string*) | The formId passed to the wrapping [Form](Form.md) component. |
| fieldId | (*string*) | The fieldId passed to the component. |
| isFormValid | (*boolean*) | Indicates whether the whole form is valid or not. |
| updateField | (*Function*) | Updates the field data.<br>Accepts an object with `isEnabled`, `isRequired`, `isTouched`, `isValid` and `value`. Each value is optional though. |
| updateState | (*Function*) | Updates the form-scoped state.<br>Takes an object of new state keys and merges those with the existing state. |


## Example
```javascript
import { Field } from 'react-controlled-form'

function Input({ fieldId }) {
  const renderInput = ({ value, updateField }) => (
    <input value={value} onChange={e => updateField({ value: e.target.value })} />
  )

  return (
    <Field 
      fieldId={fieldId} 
      render={renderInput} 
    />
  )
}

// Usage
<Input fieldId="firstname" />
<Input fieldId="lastname" />
```

## Tips & Tricks

* To provide maximum accessibility, it is recommended to utilize the default DOM attributes such as `disabled` and `required`. We may use ` isEnabled` and `isRequired` to set those attributes respectively.
* Using `onInput` instead of `onChange`, `onKeyPress` or `onKeyDown` ensures that the event has the most recent value.
* Field validation can easily be done within `onChange`. We can then set `isValid` within `updateField` to update the state. Altering the visual appearance depending on `isValid` helps to tell the user if the field is actually valid or not. Additional, we could show an error message if it's invalid.
