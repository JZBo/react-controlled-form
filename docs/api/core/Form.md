# Form

The Form component is used to declare a new form context. Every form field within its children is automatically assigned to it.

## Props
| Prop | Type | Default | Description |
| --- | --- | --- | --- |
| formId | (*string*) | |  A unique identifier which is used to save the whole form data in our Redux store. |
| initialFields | (*Object?*) | | Defines initial field data. Every field value is optional. |
| enableDefault | (*boolean?*) | false | If set, `event.preventDefault` will **not** be called on form submit. |
| onChange | (*Function?*) | | A function that is triggered on **every** change.<br>It's first parameter matches the [callback shape](#callback-shape). It additionally also receives `previousData` and `previousState` which contain the field data and form state before the change happened. |
| onSubmit | (*Function?*) | | A function that is triggered if the form gets submitted.<br>It's first parameter matches the [callback shape](#callback-shape). |
| validate | (*Function?*) | | A function to perform additional form validation.<br>It's first parameter matches the [callback shape](#callback-shape). |

#### Callback Shape
```
FieldData = {
  value: any,
  isEnabled: boolean,
  isRequired: boolean,
  isTouched: boolean,
  isValid: boolean
}

CallbackShape = {
  data: { [fieldId: string]: FieldData },
  state: { [key: string]: any },

  updateField: Function(fieldId: string, data: FieldData),
  updateState: Function(newState: Object),
  resetForm: Function
}
```

## Example
```javascript
import { Form } from 'react-controlled-form'

const initialFields = {
  country: {
    isRequired: true,
    isValid: false
  },
  language: {
    isEnabled: false
  }
}

function onChange({ updateField, data }) {
  // enable the language field and set 'English' as default
  // as soon as country has been set
  if (data.country.isTouched && data.country.value) {
    updateField('language', {
      value: 'English',
      isEnabled: true
    })
  }
}

export default () => (
  <Form formId="contact" initialFields={initialFields} render={() => (
    <form></form>
  )} />
)
```
