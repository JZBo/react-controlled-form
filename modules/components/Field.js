/* @flow */
import { Component } from 'react'
import PropTypes from 'prop-types'
import { getContext, compose } from 'recompose'
import { connect } from 'react-redux'

import { mapStateToProps, mapDispatchToProps } from '../mapping/field'

import type { Field as FieldType } from '../../types/Field'

type FieldProps = {
  // public API
  fieldId: string,
  initialData: FieldType,
  render: Function,

  // passed via Redux / context
  formId: string,
  data: FieldType,
  state: Object,
  initField: Function,
  updateField: Function,
  updateState: Function,
  isFormValid: boolean,
}

class Field extends Component {
  constructor(props, context) {
    super(props, context)

    const { initField, initialData = {} } = props
    initField(initialData)
  }

  props: FieldProps

  render() {
    const {
      data: { value = '', isEnabled, isRequired, isValid, isTouched },
      state,
      render,
      updateField,
      updateState,
      formId,
      fieldId,
      isFormValid,
    } = this.props

    return render({
      value,
      isEnabled,
      isRequired,
      isValid,
      isTouched,
      state,
      updateField,
      updateState,
      formId,
      fieldId,
      isFormValid,
    })
  }
}

export default compose(
  getContext({ formId: PropTypes.string, isFormValid: PropTypes.func }),
  connect(mapStateToProps, mapDispatchToProps)
)(Field)
