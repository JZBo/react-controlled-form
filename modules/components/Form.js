/* @flow */
import { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { shallowEqual } from 'recompose'

import objectReduce from 'fast-loops/lib/objectReduce'
import arrayEach from 'fast-loops/lib/arrayEach'

import { mapStateToProps, mapDispatchToProps } from '../mapping/form'

import type { Field, Data } from '../../types/Field'

type FormProps = {
  render: Function,
  formId: string,
  initialFields?: Data,
  initialState?: Object,
  validate?: Function,
  onChange?: Function,

  data: Field,
  state: Object,
  initForm: Function,
  resetForm: Function,
  updateField: Function,
  updateState: Function,
}

class Form extends Component {
  static childContextTypes = {
    formId: PropTypes.string.isRequired,
    subscribeToReinit: PropTypes.func,
  }

  constructor(props, context) {
    super(props, context)

    const { initForm, initialFields, initialState } = props
    initForm(initialFields, initialState)

    this.initialized = false
    this.reinitListeners = []
  }

  getChildContext() {
    return {
      formId: this.props.formId,
      subscribeToReinit: this._subscribeToReinit,
    }
  }

  /* eslint-disable consistent-return */
  componentWillReceiveProps(newProps) {
    const {
      data,
      state,
      initialFields = {},
      initialState = {},
      updateField,
      updateState,
      onChange,
    } = this.props

    const newInitialFields = newProps.initialFields || {}
    const newInitialState = newProps.initialState || {}

    // enabling re-initialisation
    if (
      !shallowEqual(initialFields, newInitialFields) ||
      !shallowEqual(initialState, newInitialState)
    ) {
      this.props.initForm(
        {
          ...initialFields,
          ...newInitialFields,
        },
        {
          ...initialState,
          ...newInitialState,
        }
      )

      arrayEach(this.reinitListeners, callback => callback())
    }

    if (
      onChange &&
      this.initialized &&
      (!shallowEqual(data, newProps.data) ||
        !shallowEqual(state, newProps.state))
    ) {
      onChange({
        reset: this.reset,
        validate: this.validate,
        data: newProps.data,
        previousData: data,
        state: newProps.state,
        previousState: state,
        updateField,
        updateState,
      })
    }

    this.initialized = true
  }
  /* eslint-enable */

  reset = () => {
    this.props.resetForm()
  }

  validate = () => {
    const { data, state, validate } = this.props

    return objectReduce(
      data,
      (isFormValid, { isValid }) => isValid && isFormValid,
      validate ? validate(data, state) : true
    )
  }

  _subscribeToReinit = callback => {
    this.reinitListeners.push(callback)

    return () =>
      this.reinitListeners.splice(this.reinitListeners.indexOf(callback), 1)
  }

  props: FormProps
  initialized: boolean
  reinitListeners: Array<Function>

  render() {
    const { formId, data, state, updateField, updateState, render } = this.props

    if (!this.initialized) {
      // quick escape as the initial rerender will already be triggered
      // TODO: it's pretty ugly to require a full rerender, maybe we can fix that
      return null
    }

    return render({
      reset: this.reset,
      validate: this.validate,
      updateField,
      updateState,
      formId,
      data,
      state,
    })
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Form)
