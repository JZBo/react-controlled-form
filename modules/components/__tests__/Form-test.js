import React from 'react'
import { createStore, combineReducers } from 'redux'

import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import Form from '../Form'
import Field from '../Field'
import { formReducer } from '../../index'

describe('Rendering a Form component', () => {
  it('should initialize the form data', () => {
    const store = createStore(
      combineReducers({
        ...formReducer,
      })
    )

    const wrapper = mount(
      <Form
        formId="name"
        initialFields={{
          firstname: {
            value: 'John',
          },
          lastname: { value: 'Doe' },
        }}
        render={({ data }) => (
          <div>
            Hello {data.firstname.value} {data.lastname.value}
          </div>
        )}
      />,
      {
        context: {
          store,
        },
      }
    )

    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should initialize the form data form inner fields', () => {
    const store = createStore(
      combineReducers({
        ...formReducer,
      })
    )

    const wrapper = mount(
      <Form
        formId="name"
        render={({ data }) => (
          <div>
            Hello {(data.firstname && data.firstname.value) || ''}.
            <Field
              fieldId="firstname"
              initialData={{ value: 'John' }}
              render={({ value }) => <input value={value} />}
            />
          </div>
        )}
      />,
      {
        context: {
          store,
        },
      }
    )

    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
