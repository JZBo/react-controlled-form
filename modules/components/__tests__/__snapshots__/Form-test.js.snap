// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering a Form component should initialize the form data 1`] = `
<Connect(Form)
  formId="name"
  initialFields={
    Object {
      "firstname": Object {
        "value": "John",
      },
      "lastname": Object {
        "value": "Doe",
      },
    }
  }
  render={[Function]}
>
  <Form
    data={
      Object {
        "firstname": Object {
          "value": "John",
        },
        "lastname": Object {
          "value": "Doe",
        },
      }
    }
    formId="name"
    initForm={[Function]}
    initialFields={
      Object {
        "firstname": Object {
          "value": "John",
        },
        "lastname": Object {
          "value": "Doe",
        },
      }
    }
    render={[Function]}
    resetForm={[Function]}
    state={Object {}}
    updateField={[Function]}
    updateState={[Function]}
  >
    <div>
      Hello 
      John
       
      Doe
    </div>
  </Form>
</Connect(Form)>
`;

exports[`Rendering a Form component should initialize the form data form inner fields 1`] = `
<Connect(Form)
  formId="name"
  render={[Function]}
>
  <Form
    data={
      Object {
        "firstname": Object {
          "isEnabled": true,
          "isRequired": false,
          "isTouched": false,
          "isValid": true,
          "value": "John",
        },
      }
    }
    formId="name"
    initForm={[Function]}
    render={[Function]}
    resetForm={[Function]}
    state={Object {}}
    updateField={[Function]}
    updateState={[Function]}
  >
    <div>
      Hello 
      John
      .
      <getContext(Connect(Field))
        fieldId="firstname"
        initialData={
          Object {
            "value": "John",
          }
        }
        render={[Function]}
      >
        <Connect(Field)
          fieldId="firstname"
          formId="name"
          initialData={
            Object {
              "value": "John",
            }
          }
          render={[Function]}
          subscribeToReinit={[Function]}
        >
          <Field
            data={
              Object {
                "isEnabled": true,
                "isRequired": false,
                "isTouched": false,
                "isValid": true,
                "value": "John",
              }
            }
            fieldId="firstname"
            formId="name"
            initField={[Function]}
            initialData={
              Object {
                "value": "John",
              }
            }
            render={[Function]}
            state={Object {}}
            subscribeToReinit={[Function]}
            updateField={[Function]}
            updateState={[Function]}
          >
            <input
              value="John"
            />
          </Field>
        </Connect(Field)>
      </getContext(Connect(Field))>
    </div>
  </Form>
</Connect(Form)>
`;
